src/hotspot/share/utilities/enumIterator.hpp

```diff
@@ -0,0 +1,214 @@
+// Not clear what iteration should mean in this case.
+// Don't bother trying to figure this out.
+//
+//
+// EnumIterationTraits -- defines the static range of all possible values of the enum.
+// EnumRange -- defines the range of *one specific* iteration loop.
+// EnumIterator -- the current point in the iteration loop.
+
+// Example (see vmSymbols.hpp/cpp)
+//
+// ENUMERATOR_RANGE(vmSymbolID, vmSymbolID::FIRST_SID, vmSymbolID::LAST_SID)
+// constexpr EnumRange<vmSymbolID> vmSymbolsRange;
+// using vmSymbolsIterator = EnumIterator<vmSymbolID>;
+//
+// /* Without range-based for, allowed */
+// for (vmSymbolsIterator it = vmSymbolsRange.begin(); it != vmSymbolsRange.end(); ++it) {
+//  vmSymbolID index = *it; ....
+// }
+//
+// /* With range-base for, not allowed by HotSpot coding style yet */
+// for (vmSymbolID index : vmSymbolsRange) {
+//    ....
+// }
+
+// Traits type supporting iteration over the enumerators of T.
+// Specializations must provide static const data members named
+// "_first" and "_last", whose values are the smallest / largest
+// (resp.) enumerator values for T. For iteration, the enumerators of
+// T must have sequential values in that range.
+template<typename T> struct EnumeratorRange;
+
+// Specialize EnumeratorRange<T>.
+#define ENUMERATOR_RANGE(T, First, Last)        \
+  template<> struct EnumeratorRange<T> {        \
+    static constexpr T _first = First;          \
+    static constexpr T _last = Last;            \
+  };
+
+template<typename T>
+class EnumIterationTraits : AllStatic {
+  using RangeType = EnumeratorRange<T>;
+
+public:
+  // The underlying type for T.
+  using Underlying = std::underlying_type_t<T>;
+
+  // The first enumerator of T.
+  static constexpr T _first = RangeType::_first;
+
+  // The last enumerator of T.
+  static constexpr T _last = RangeType::_last;
+
+  static_assert(static_cast<Underlying>(_last) <
+                std::numeric_limits<Underlying>::max(),
+                "No one-past-the-end value for enum");
+
+  // The value of the first enumerator of T.
+  static constexpr Underlying _start = static_cast<Underlying>(_first);
+
+  // The one-past-the-end value for T.
+  static constexpr Underlying _end = static_cast<Underlying>(_last) + 1;
+};
+
+template<typename T>
+class EnumIterator {
+  using Traits = EnumIterationTraits<T>;
+
+  using Underlying = typename Traits::Underlying;
+  Underlying _value;
+
+  constexpr void assert_in_bounds() const {
+    assert(_value < Traits::_end, "beyond the end");
+  }
+
+public:
+  // Return a beyond-the-end iterator.
+  constexpr EnumIterator() : _value(Traits::_end) {}
+
+  // Return an iterator with the indicated value.
+  constexpr explicit EnumIterator(T value) :
+    _value(static_cast<Underlying>(value))
+  {
+    assert(_value >= Traits::_start, "out of range");
+    assert(_value <= Traits::_end, "out of range");
+  }
+
+  // True if the enumerators designate the same value.
```

Comment:
"True if the iterators designate the same enumeration value."  (Sorry about that.)  Similarly below for operator!=.